pkg/agent/api/delegatedidentity/v1/service.go

```diff
@@ -0,0 +1,228 @@
+	resp, err := composeX509SVIDBySelectors(update)
+	if err != nil {
+		log.WithError(err).Error("Could not serialize X.509 SVID response")
+		return status.Error(codes.Unavailable, "could not serialize response")
+	}
+
+	if err := stream.Send(resp); err != nil {
+		log.WithError(err).Error("Failed to send X.509 SVID response")
+		return err
+	}
+
+	return nil
+}
+
+func composeX509SVIDBySelectors(update *cache.WorkloadUpdate) (*delegatedidentityv1.SubscribeToX509SVIDsResponse, error) {
+	resp := new(delegatedidentityv1.SubscribeToX509SVIDsResponse)
+	resp.X509Svids = []*delegatedidentityv1.X509SVIDWithKey{}
+
+	for td, _ := range update.FederatedBundles {
+		resp.FederatesWith = append(resp.FederatesWith, td.IDString())
+	}
+
+	for _, identity := range update.Identities {
+		// TODO:  it doesn't work and always prints "false"
+		//fmt.Printf("processing id %s, %t\n", identity.Entry.SpiffeId, identity.Entry.Admin)
+		// Do not send admin nor downstream SVIDs to the caller
+		if identity.Entry.Admin || identity.Entry.Downstream {
+			continue
+		}
+
+		id, _ := idutil.IDProtoFromString(identity.Entry.SpiffeId)
+
+		keyData, err := x509.MarshalPKCS8PrivateKey(identity.PrivateKey)
+		if err != nil {
+			return nil, fmt.Errorf("marshal key for %v: %w", id, err)
+		}
+
+		svid := &delegatedidentityv1.X509SVIDWithKey{
+			X509Svid: &types.X509SVID{
+				Id:        id,
+				CertChain: x509util.RawCertsFromCertificates(identity.SVID),
+				// TODO: what if SVIDs 0 doesn't exist?
+				ExpiresAt: identity.SVID[0].NotAfter.Unix(),
+			},
+			X509SvidKey: keyData,
+		}
+		resp.X509Svids = append(resp.X509Svids, svid)
+	}
+	return resp, nil
+}
+
+func (s *Service) SubscribeToX509Bundles(req *delegatedidentityv1.SubscribeToX509BundlesRequest, stream delegatedidentityv1.DelegatedIdentity_SubscribeToX509BundlesServer) error {
+	ctx := stream.Context()
+	log := rpccontext.Logger(ctx)
+
+	authorized, err := s.isCallerAuthorized(ctx)
+
+	if err != nil {
+		log.WithError(err).Error("Workload attestation failed")
+		return status.Error(codes.Internal, "workload attestation failed")
+	}
+
+	if !authorized {
+		return status.Error(codes.PermissionDenied, "no authorized")
+	}
+
+	subscriber := s.manager.SubscribeToBundleChanges()
+
+	// send initial update....
+	for td, bundle := range subscriber.Value() {
+		resp := &delegatedidentityv1.SubscribeToX509BundlesResponse{
+			CaCertificates: map[string][]byte{td.IDString(): marshalBundle(bundle.RootCAs())},
+		}
+
+		if err := stream.Send(resp); err != nil {
+			return err
+		}
+	}
+
+	for {
+		select {
+		case <-subscriber.Changes():
+			for td, bundle := range subscriber.Next() {
+				resp := &delegatedidentityv1.SubscribeToX509BundlesResponse{
+					CaCertificates: map[string][]byte{td.IDString(): marshalBundle(bundle.RootCAs())},
+				}
+
+				if err := stream.Send(resp); err != nil {
+					return err
+				}
+			}
```

Comment:
I think the behavior we discussed is that the CaCertificates field should include the bundles of _all_ federated trust domains being managed by the agent. Instead, this code looks like we are sending one response per federated trust domain?

https://github.com/spiffe/spire-api-sdk/blob/main/proto/spire/api/agent/delegatedidentity/v1/delegatedidentity.proto#L53-L60

Reply:
Yeah, you are right... now is going to send all bundles in only one response :D